Add basic EVP_KEYMGMT API and libcrypto <-> provider interface

levitte opened the pull request:
The idea with the key management "operation" is to support the
following set of functionality:

- Key domain parameter generation
- Key domain parameter import
- Key domain parameter export

- Key generation
- Key import
- Key export
- Key loading (HSM / hidden key support)

With that set of function, we can support handling domain parameters
on one provider, key handling on another, and key usage on a third,
with transparent export / import of applicable data.  Of course, if a
provider doesn't offer export / import functionality, then all
operations surrounding a key must be performed with the same
provider.

This method also avoids having to do anything special with legacy
assignment of libcrypto key structures, i.e. EVP_PKEY_assign_RSA().
They will simply be used as keys to be exported from whenever they are
used with provider based operations.

This change only adds the EVP_KEYMGMT API and the libcrypto <->
provider interface.  Further changes will integrate them into existing
libcrypto functionality.

levitte committed:
Add basic EVP_KEYMGMT API and libcrypto <-> provider interface

The idea with the key management "operation" is to support the
following set of functionality:

- Key domain parameter generation
- Key domain parameter import
- Key domain parameter export

- Key generation
- Key import
- Key export
- Key loading (HSM / hidden key support)

With that set of function, we can support handling domain parameters
on one provider, key handling on another, and key usage on a third,
with transparent export / import of applicable data.  Of course, if a
provider doesn't offer export / import functionality, then all
operations surrounding a key must be performed with the same
provider.

This method also avoids having to do anything special with legacy
assignment of libcrypto key structures, i.e. EVP_PKEY_assign_RSA().
They will simply be used as keys to be exported from whenever they are
used with provider based operations.

This change only adds the EVP_KEYMGMT API and the libcrypto <->
provider interface.  Further changes will integrate them into existing
libcrypto functionality.

levitte commented:
Do not that this is a *very early draft*.  The primary purpose for submitting this PR this early is to expose they ideas and get early feedback on the line of thoughts.

levitte commented:
> - Key loading (HSM / hidden key support)

This is highly experimental for now, but does provide the necessary support for getting keys through provider OSSL_STORE backends.  More on that in a separate PR.

paulidale commented:
I really want to review this in more detail.  I didn't get to it today and am too tired to think too hard this evening.

levitte committed:
Add evp_keymgmt_export_to_provider(), for key transfer between providers

This function is used to transport private key materia from whatever
is already attached to the EVP_PKEY to the new provider, using key
data export and import functionality.

If a legacy lower level key has been assigned to the EVP_PKEY, we use
its data to create a provider side key, and thereby have a bridge
between old style public key types and the EVP_PKEY on providers.

If successful, this function returns a reference to the appropriate
provider side data for the key.

This can be used by any operation that wants to use this key.

petrovr commented:
Key Derivation?

I'm not sure about operations "Key loading" vs "Key import". Second is defined for private and public. What will load first one?

petrovr commented:
Perhaps API could be based on basic(elementary/atomic) operations 
defined in "key management protocol".
There is some standard protocols but I prefer experts to propose which 
one to use.

Roumen

levitte commented:
> Key Derivation?

Different operation, see #9266...  but yes, it has occured to @mattcaswell and me that key exchange result in a key as well, that wasn't lost on us.  There will be some form of integration eventually, I'm sure, but not *now* and not in *this PR*.  Let's take it a step at a time, yeah?

levitte commented:
> Perhaps API could be based on basic(elementary/atomic) operations defined in "key management protocol".

Not sure if that would make things better or not.  Maybe...

levitte commented:
> I'm not sure about operations "Key loading" vs "Key import".

The key import function takes key material from the input `OSSL_PARAM` array and creates a key from that input (usually a bunch of big numbers).
The key loading function is aimed at HSMs, and gives the option to "load" an HSM protected key given some kind of identity.  The identity isn't defined at all for the moment, and I'm thinking of it as highly provider specific, i.e. unstandardised.  My current thinking is that it can work together with a provider OSSL_STORE implementation that returns such identities instead of EVP_PKEYs, and leaves it to libcrypto to actually "load" the key given such an identity.  I've an unsubmitted branch where I currently experiment with this.

levitte commented:
Documentation added

levitte committed:
Adapt int_ctx_new() to use with providers

This affects all its callers: EVP_PKEY_CTX_new(), EVP_PKEY_CTX_new_id().
They are now possible to called with "zero" values, i.e.:

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(NULL, NULL);

or

    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(0, NULL);

This is suitable for provider use, as the key functionality is tied
with its keys, and the operation time is determined by the init
functions the EVP_PKEY_CTX is used with.

petrovr commented:
Richard Levitte wrote:
>> I'm not sure about operations "Key loading" vs "Key import".
> 
> The key import function takes key material from the input `OSSL_PARAM` array and creates a key from that input (usually a bunch of big numbers).

Interesting.
Working with different cryptographic libraries I note that common part 
is exchange of DER encoded data (key, cert). So instead ASN1 encoding 
you prefer data encoded as OSSL_PARAM, right?


> The key loading function is aimed at HSMs, and gives the option to "load" an HSM protected key given some kind of identity.  The identity isn't defined at all for the moment, and I'm thinking of it as highly provider specific, i.e. unstandardised.  
This cannot be restricted to "hardware" only.  Software "key-store" may 
not allow private part to be exported.


> My current thinking is that it can work together with a provider OSSL_STORE implementation that returns such identities instead of EVP_PKEYs, and leaves it to libcrypto to actually "load" the key given such an identity.  I've an unsubmitted branch where I currently experiment with this.
> 

Right now I'm not sure what you mean by provider and what is difference 
to engine. Shall we consider engine as simple provider?

Regards,
Roumen

levitte commented:
I think you need to read up in https://www.openssl.org/docs/OpenSSL300Design.html and https://www.openssl.org/docs/OpenSSLStrategicArchitecture.html.  The former is what were working toward, and is the reason you're seeing so much moving around right now.

petrovr commented:
Richard Levitte wrote:
>The identity isn't defined at all for the moment, and I'm thinking of it as highly provider specific, i.e. unstandardised.

In some implementation "identity" is X.509 certificate(a) and private 
key. For other is pair public(b) and private key. Private key is 
accessed indirectly by (a) or by (b).
May be you want to use another kind of keys . But this will require 
delegation of cryptographic operations to identity or "identity provider".

Roumen

levitte commented:
> Working with different cryptographic libraries I note that common part is exchange of DER encoded data (key, cert). So instead ASN1 encoding you prefer data encoded as OSSL_PARAM, right?

We have chosen the OSSL_PARAM array as a data transfer medium between libcrypto and provider, yes.  We wanted something that isn't tied to any particular form of serialization.  I've realized after the fact that they have certain similarities with PKCS#11's attribute templates.

levitte commented:
> > The key loading function is aimed at HSMs, and gives the option to "load" an HSM protected key given some kind of identity.  The identity isn't defined at all for the moment, and I'm thinking of it as highly provider specific, i.e. unstandardised.
>
> This cannot be restricted to "hardware" only.  Software "key-store" may not allow private part to be exported.

Correct, I use "HSM" loosly, i.e. for something that keeps keys and other sensitive data for themselves.

levitte committed:
Add evp_keymgmt_clear_pkey_cache() and use it

This function clears the cache of provider key references, and is used
in evp_keymgmt_export_to_provider() when the internal key is dirty, as
well as by EVP_PKEY_free_it().

levitte committed:
Add an internal API to access the KEYMGMT provider functions

levitte commented:
This is not a draft any more

mattcaswell commented:
"make doc-nits" is complaining.

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+
+    /*
+     * If there are no cached provider keys, try to export a legacy key
+     * to the provider.
```
Should we do this whenever there is legacy key available rather than if there is no cached keys. Otherwise it seems this function will fail in the scenario where we have have a legacy key, we have exported it to a provider, but that provider does not provide an import implementation. It seems odd to do the import/export thing if we already have the information we need in a legacy key.

levitte replied:
Hmmm, good point.  That would make for one less OSSL_PARAM array transfer too.

levitte replied:
Did quite a bit of rewrite, I hope I fixed this.

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+     * and export the imported data to the new provider.
+     * If it's impossible to import from the cached provider, give up
+     * (transfering keys between providers can't be supported in that
+     * case)
```
I'm getting kind of confused about which direction we are talking about with export/import and from which perspective we are talking about it. i.e. does "import" mean import into a provider, or do we "import" a key from a provider into libcrypto. Above you talk about "export...to the new provider", but below you call the "importkey_priv" function.

levitte replied:
Hahaha, well, who's importing and who's exporting, that's the question, right?
So if the U.K. exports their Christmas Pudding, does Sweden import it or export it?
So on the same line of thinking, if libcrypto is exporting a key to a provider, is the provider importing or exporting it?

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+        void *data = NULL;
+        void *provctx = ossl_provider_ctx(EVP_KEYMGMT_provider(keymgmt));
+
+        paramdefs = pk->pkeys[0].keymgmt->exportkey_priv_types();
```
We seem to be talking specifically about private keys here. What about public keys?

levitte replied:
Ah, good question.  The "traditional" OpenSSL setup is that the private key *always* also holds the public key, while the public key doesn't hold the private key.  Considering we also usually pass the numbers that were used to calculate the keys themselves, it's really rather pointless *not* to include the public bits with the private bits, but maybe this is something we'd want to rethink?

Otherwise, maybe we should have "key" and "key_pub" rather than "key_priv" and "key_pub".

I'm open for alternatives

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+         */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
+        pk->pkeys[0].provkey = keymgmt->importkey_priv(provctx, params);
```
If the returned provkey is NULL then shouldn't we free keymgmt? Or is the intent to actually cache the failure?

levitte replied:
Ah, no, I simply forgot.  Thanks for being eagle-eyed :-)

levitte replied:
Fixed, The new array element is now only set if we actually got a provider key

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+         * new provider and hope to get a key back.
+         */
+        EVP_KEYMGMT_up_ref(keymgmt);
+        pk->pkeys[0].keymgmt = keymgmt;
```
Shouldn't this be using `i` here rather than 0?

levitte replied:
Ah, yes, that's right.

levitte replied:
Fixed.

mattcaswell commented on crypto/evp/evp_locl.h:
```diff
@@ -60,6 +60,36 @@ struct evp_kdf_ctx_st {
+    CRYPTO_RWLOCK *lock;
+
+    /* Domain parameter routines */
+    OSSL_OP_keymgmt_importdomain_fn *importdomain;
```
I find the use of "domain" here surprising. Elsewhere when talking about parameters we use the abbreviation "param"

levitte replied:
From reading up on the nomenclature for dh and dss, I see "domain parameters" used everywhere.  That's the basis for my choice, and also a wish to not have that too easily mixed up with the standard `set_params` / `get_params` calls.

levitte replied:
(standard `set_params` / `get_params` calls don't create objects...)

mattcaswell replied:
I suppose "domainparams" might work, although it is a little long.

mattcaswell replied:
Or "algparams"

levitte replied:
"domparams"?

mattcaswell replied:
domparams is fine by me.

levitte replied:
Ok, will fix

mattcaswell replied:
This still seems to be outstanding

levitte replied:
Ah, right

levitte replied:
Done

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+{
+    size_t i;
+
+    if (pk != NULL)
```
Although technically the body of this "if" is a single statement, I normally advise contributors to enclose such multi-line bodies inside {}

levitte replied:
Okie, will do

levitte replied:
Done

mattcaswell commented on crypto/evp/keymgmt_meth.c:
```diff
@@ -0,0 +1,197 @@
+            break;
+        }
+    }
+    keymgmt->prov = prov;
```
Should we have some consistency checking here? For example does it ever make sense to not have a freekey function when you have a loadkey/genkey function? Or to not have importkey_pub, when importkey_pub_types is non NULL? etc

levitte replied:
Good point.  Will add a check

levitte replied:
Done

mattcaswell commented on crypto/evp/pmeth_lib.c:
```diff
@@ -103,8 +103,17 @@ const EVP_PKEY_METHOD *EVP_PKEY_meth_find(int type)
+    const EVP_PKEY_METHOD *pmeth = NULL;
 
+    /*
+     * For new style use, the context is bound to the algo implementation
```
"new style" won't make much sense in the future, when it isn't "new" any more.

levitte replied:
Good point.  What do you usually do when you want to say something like this?  (or should we go for "pink nirvana"?)

mattcaswell replied:
How about "When using providers" or something similar?

levitte replied:
Thanks.  Will do

levitte replied:
Done

mattcaswell commented on doc/internal/man3/evp_keymgmt_export_to_provider.pod:
```diff
@@ -0,0 +1,55 @@
+
+=head1 HISTORY
+
+The functions described here were all added in OpenSSL 3.0.
```
I am dubious about the value of the HISTORY section in internal docs. In the public API it is useful to know when a function was introduced if you are wanting to target multiple versions of OpenSSL. For internal functions that isn't the case (you're only ever interested in the "current" version), and even if a function was present in an old release its purpose or signature might have changed over time (which is again not the case with public API functions).

levitte replied:
Hmmm, good point.  I kinda had that out of habit.  Will remove

levitte replied:
Done

mattcaswell commented on doc/internal/man3/evp_keymgmt_freekey.pod:
```diff
@@ -0,0 +1,137 @@
+All these functions are helpers to call the provider's corresponding
+function.
+
+evp_keymgmt_importdomain() calls the method's importdomain() function.
```
Should we be documenting what "importdomain" actually *does*? Similarly with the other lines below.

levitte replied:
Yeah, we should.  I *also* think we should document the dispatch functions, I just dunno how, or rather where.  (I'm trying to document in the header file at the very least, something we generally suck at, I might add...)

mattcaswell replied:
Whether in this PR or another, we're going to need to document all of these provider functions in the man pages. A provider author should be able to find all the information they need in the man pages, without have to read the source.

levitte replied:
Agreed.  For the `core_numbers.h` stuff, that will be another PR.  I have ideas on that...

mattcaswell commented on doc/man3/EVP_KEYMGMT.pod:
```diff
@@ -0,0 +1,84 @@
+
+ typedef struct evp_keymgmt_st EVP_KEYMGMT;
+
+ EVP_KEYMGMT *EVP_KEYMGMT_fetch(OPENSSL_CTX *ctx, const char *algorithm,
```
My idea was to document all "fetch" functions in the same place. See the docs for `EVP_MD_fetch` and `EVP_CIPHER_fetch`

levitte replied:
Yeah, I noticed...  and quite frankly, I think that's not a good idea, as it separates one KEYMGMT function from the rest, and I'm thinking the same about the other fetch manuals.  I've just been too preoccupied to comment on that earlier.

What I *do* think is also that we make too little use of man7.  We could have a page about the ins and outs of fetching there, i.e. much of the prose in EVP_MD_fetch.pod, and then refer back to it from any page that describes the actual fetching function.

mattcaswell replied:
I could live with that.

levitte replied:
I'll start that in another PR

mattcaswell replied:
Just a reminder to start this other PR

levitte replied:
#9409

mattcaswell commented on doc/man3/EVP_KEYMGMT.pod:
```diff
@@ -0,0 +1,84 @@
+
+EVP_KEYMGMT_fetch() looks for an algorithm within the provider that
+has been loaded into the B<OPENSSL_CTX> given by I<ctx>, having the
+name given by I<algorithm> and the properties given by I<properties>.
```
OPENSSL_CTX can be NULL in which case you get the default OPENSSL_CTX.

levitte replied:
I should probably add a `L<OPENSSL_CTX(3)>` in SEE ALSO, then.  Or do we need to repeat the fact about NULL in every spot we mention `OPENSSL_CTX`?

levitte replied:
Link added

mattcaswell reviewed (changes requested):
We should have some tests for this code.

levitte commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+        if (keymgmt == pk->pkeys[i].keymgmt)
+            return pk->pkeys[i].provkey;
+    }
+
```
I just realised nothing is done about the case where all the slots of `pk->pkeys` are filled in and none of them matched the given keymgmt, i.e. we're not dropping any single entry from the cache...  Needs to be fixed.

levitte commented:
> We should have some tests for this code.

Ah, in `test/evp_extra_test.c`?  (I noticed that `EVP_MD_fetch` has a few tests there)

Sure, I will, if you add one for `EVP_CIPHER_fetch` :wink:

levitte commented:
> "make doc-nits" is complaining.

Fixed

paulidale commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,289 @@
+    const OSSL_PARAM *p;
+    OSSL_PARAM *params, *q;
+
+    for (cnt = 0, p = paramdefs; p->key != NULL; p++, cnt++)
```
This looks off by one low?

levitte replied:
I don't follow...

paulidale replied:
Consider, the trivial case: `paramdefs = { OSSL_PARAM_END }`.
After the loop finishes, `cnt` will be zero.

The allocated array (of zero length) would not have enough space to be terminated but it is terminated.

Changing the `cnt = 0` to `cnt = 1` should fix this.

levitte replied:
Okie, will look again

levitte replied:
Fixed

paulidale commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+int evp_keymgmt_exportkey_priv(const EVP_KEYMGMT *keymgmt,
+                               void *provkey, OSSL_PARAM params[])
+{
+    return keymgmt->exportkey_priv(provkey, params);
```
Should these function pointers be NULL checked?

levitte replied:
Would it be enough if they were checked by the caller?  Or I dunno...  this is an internal function, so my attitude is a bit more lax on the matter...

paulidale replied:
Since they are not public, yes it would be sufficient for the caller to check.
However, the overhead for doing so here is minimal and I'm not expecting these calls to be performance critical.

paulidale commented on crypto/evp/keymgmt_meth.c:
```diff
@@ -0,0 +1,197 @@
+
+    for (; fns->function_id != 0; fns++) {
+        switch (fns->function_id) {
+        case OSSL_FUNC_KEYMGMT_IMPORTDOMAIN:
```
Not for this PR but it might be worth having a table based function for this kind of code?
```
    { OSSL_FUNC_KEYMGMT_IMPORTDOMAIN, &keymgmt->importdomain, OSSL_get_OP_keymgmt_importdomain }
```

levitte replied:
I'll just say this: ISO C90.

paulidale replied:
?????

Not allowed to take the address of fields in structures?
Or did I miss variable arguments?

levitte replied:
`&keymgmt->importdomain`

That address will have to be calculated in run time, i.e. the compiler can't determine it.

paulidale replied:
Good point.  I'm not seeing properly today :(

paulidale commented on doc/internal/man3/evp_keymgmt_export_to_provider.pod:
```diff
@@ -0,0 +1,51 @@
+
+If I<pk> has an assigned internal key, a check is done to see if any
+of its key material has changed since last export.
+If it has (the is_dirty() method returned 1), the cache of already
```
The parenthesised part could go at the end of the previous sentence.

levitte replied:
Okie

paulidale reviewed (changes requested):
Unless I'm mistaken (which has been common today).

levitte commented:
There are quite a few conversations going on here, I'd like to nudge y'all to respond

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+            return NULL;
+
+        if (pk->ameth->is_dirty(pk))
+            evp_keymgmt_clear_pkey_cache(pk);
```
Do we need to clear the dirty flag here?

levitte replied:
Done

levitte replied:
(although, the right spot is after the `export_to` call)

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+
+                if (provkey != NULL) {
+                    break;
+                }
```
Nit: Remove the {}

levitte replied:
Will do

levitte replied:
Done

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+     * have to think about a cache aging scheme, though, if |i| indexes
+     * outside the array.
+     */
+    j = ossl_assert(i < OSSL_NELEM(pk->pkeys));
```
I think this should just return an error rather than asserting this. It *could* happen through normal use (although unlikely) - whereas an assert should only be for "this should never happen".

levitte replied:
I'm really having that in place of a proper cache ageing mechanism, so *for now*, I've left that unimplemented with the belief that we won't hit that many exports simply because there are only so many providers available, but we *should* implement a proper ageing mechanism at some point before 3.0 is released.  I see no reason to play nice in the mean time, and should this assertion be triggered before then, we'd better hustle our buns to get that ageing mechanism in place.

mattcaswell replied:
ok

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+
+    EVP_KEYMGMT_up_ref(keymgmt);
+    pk->pkeys[i].keymgmt = keymgmt;
+    pk->pkeys[i].provkey = provkey;
```
Shouldn't there be a test for provkey being NULL somewhere around here?

levitte replied:
Why?  What will we do as a result...  apart from returning NULL?

mattcaswell replied:
Well we wouldn't up ref keymgmt or cache it in that case.

levitte replied:
Ooooh, you're right!  Thanks, eagle-eye.  Will fix.

levitte replied:
Done

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+             i < OSSL_NELEM(pk->pkeys) && pk->pkeys[i].keymgmt != NULL;
+             i++) {
+            pk->pkeys[i].keymgmt->freekey(pk->pkeys[i].provkey);
+            pk->pkeys[i].keymgmt = NULL;
```
`EVP_KEYMGMT_free(pk->pkeys[i].keymgmt)`?

levitte replied:
Ah, good point.  Will do

levitte replied:
Done

mattcaswell commented on crypto/include/internal/asn1_int.h:
```diff
@@ -63,6 +63,10 @@ struct evp_pkey_asn1_method_st {
     int (*get_pub_key) (const EVP_PKEY *pk, unsigned char *pub, size_t *len);
+
+    /* Exports to providers */
+    int (*is_dirty) (const EVP_PKEY *pk);
```
I think we'll need something to reset the dirty flag

levitte replied:
I was thinking that it'd be a job for `export_to`, but I do see the point with making it possible to clear from the exterior.  Will do

levitte replied:
Done

mattcaswell commented on doc/internal/man3/evp_keymgmt_export_to_provider.pod:
```diff
@@ -0,0 +1,56 @@
+=head1 RETURN VALUES
+
+evp_keymgmt_export_to_provider() returns a pointer to the newly
+created provider side key, or NULL on error.
```
Or the pointer to a previously cached provider side key?

levitte replied:
Ah, right.  Will fix

levitte replied:
Done

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,307 @@
+        /*
+         * If the given keymgmt doesn't have an import function, give up
+         */
+        if (keymgmt->importkey_priv == NULL)
```
So I'm trying to work out whether we actually *need* the import/export key_pub variants since we never seem to call them. Is it enough just to import/export a key, with the private fields being optional? Alternatively do we need to detect whether a key is private or public and then call the right function?

levitte replied:
Well, I guess it comes down to certainty...  when do you want to know that you have the right set of key material for your operation?  Typically, encrypt and verify requires a private key while decrypt and sign requires a public key.  Do you want to know what you have before engaging the operation, or is it enough that the operation comes back with a "Hey dude, this key doesn't look right"?

mattcaswell replied:
Well, if we're after certainty then calling `importkey_priv` unconditionally in this code does not seem the right thing to do since it won't always be a private key. Presumably `importkey_priv` would be expected to fail if we only passed it a public key. I'm tempted to say that its ok that its not until you attempt the operation that you hit the problem. After all that's basically what we have now - which would suggest we don't need different variants for priv vs pub.

levitte replied:
Ok, that's a good point ('cause you're right, when a `EVP_PKEY_assign_RSA` is done, we have no idea what's in the assigned legacy key).

Will fix

mattcaswell replied:
This still seems to be an issue?

levitte replied:
Thanks for the reminder.

levitte replied:
Done

petrovr commented:
Engine interface lacks functionality to get X.509 certificate for a key. As work around some engines implement "custom" control commands. More or less STORE2 API removes such limitation. 

It seems to me new API reintroduce engine limitation.

levitte commented:
> It seems to me new API reintroduce engine limitation.

The KEYMGMT API doesn't support OSSL_STORE, no.  I'm working on *another* libcrypto\<-\>provider API that supports OSSL_STORE.  The plan is that *that* API will return key references in form of identities that can be used with the `loadkey` function in this API.  This is to allow the EVP_PKEY to be created by libcrypto without too much weird entanglement in the provider.

Please be patient.

mattcaswell commented:
Looks like you need a rebase

levitte commented:
> Looks like you need a rebase

Yup, I noticed.  Especially with #9266 just being merged...

levitte commented:
Rebased and squashed, to get a clean slate.

levitte committed:
fixup! Add evp_keymgmt_export_to_provider(), for key transfer between providers

levitte committed:
fixup! Add basic EVP_KEYMGMT API and libcrypto <-> provider interface

levitte committed:
fixup! Add evp_keymgmt_clear_pkey_cache() and use it

levitte commented:
I think this is ready for final review.  There was a failure in Travis, but that turned out to be an unrelated timeout.  I've restarted that job, just for the hope of greenery :wink:

levitte committed:
fixup! Add basic EVP_KEYMGMT API and libcrypto <-> provider interface

levitte committed:
fixup! Add an internal API to access the KEYMGMT provider functions

levitte committed:
fixup! Add evp_keymgmt_export_to_provider(), for key transfer between providers

mattcaswell commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -0,0 +1,288 @@
+    } else {
+        /*
+         * Here, there is no legacy key, so we look at the already
+         * cached but there are cached provider keys, and import from
```
This sentence doesn't seem to quite make sense

levitte replied:
Rephrased.

mattcaswell commented on crypto/include/internal/asn1_int.h:
```diff
@@ -63,6 +63,10 @@ struct evp_pkey_asn1_method_st {
     int (*get_priv_key) (const EVP_PKEY *pk, unsigned char *priv, size_t *len);
     int (*get_pub_key) (const EVP_PKEY *pk, unsigned char *pub, size_t *len);
+
+    /* Exports to providers */
```
Perhaps we should add a TODO here (or somewhere else appropriate) to remind us to go and implement these for the various legacy keys that we have?

levitte replied:
Okie, added one here

mattcaswell reviewed (approved):
LGTM. A couple of minor comments which may or may not be addressed before committing. Approval counts either way.

levitte commented:
Merged.

a94a3e0d91378b5c478f687a0dbc51914d4ed497 Add basic EVP_KEYMGMT API and libcrypto <-> provider interface
70a1f7b4d7a0611508f45ef884472b9d84cbe108 Add evp_keymgmt_export_to_provider(), for key transfer between providers
d0ea49a820e02713bbe8892a333f2552da633b16 Adapt int_ctx_new() to use with providers
4cae07fef3b24bee1646f6085c52175ef3755d5b Add evp_keymgmt_clear_pkey_cache() and use it
fa9faf010460f5fe0722a65f9d496221c2f41f7b Add an internal API to access the KEYMGMT provider functions
